implemented '%' for matching pairs of '{[()]}'

DanEEStar opened the pull request:
fixes #80 

This is my biggest pull request yet.
It was pretty easy to implement, because I could reuse a lot of the existing code.
Below are some/comments question to some specific parts.

DanEEStar commented on src/TextObjects/Block.ts:
```diff
@@ -42,7 +42,7 @@ export class TextObjectBlock extends TextObject {
     }
 
-    protected findStartRange(document:TextDocument, anchor: Position): Range {
+    public findStartRange(document:TextDocument, anchor: Position): Range {
```
I had to make this functions public. I could not simply use the `apply` method of the `TextObject` because I needed the `startRange.start` and/or the `endRange.start` property...

aioutecism replied:
You are right.
Besides, It's not so bad to make these functions public.

DanEEStar commented on src/Motions/MatchPair.ts:
```diff
@@ -0,0 +1,57 @@
+
+        const document = activeTextEditor.document;
+        
+        const characterUnderCursor = document.getText(new Range(from, new Position(from.line, from.character+1)));
```
This is a pretty generic method. Should this be moved to `Utils`?

aioutecism replied:
Sure.
I think it would be more efficient to do it like this:

```ts
document.getText(new Range(
    anchor.line, anchor.character,
    anchor.line, anchor.character + 1
));
```

aioutecism replied:
However, we may not need this method since we need to find next start/end string in this line.

DanEEStar commented on src/Motions/MatchPair.ts:
```diff
@@ -0,0 +1,57 @@
+            return from;
+        }
+        
+        const textObject = MotionMatchPair.characterMap[characterUnderCursor]({isInclusive: true});
```
As written below this function needs neither isInclusive nor not isInclusive.
Maybe it would be better to have a enum with three options for the `TextObject` constructor like
```
enum TextObjectOptions {
    InnerBlock, // like isInclusive
    Block,
    StartRange, // used for the % motion
}
```
    Hearts,
    Spades
}
?

aioutecism commented:
```
%

Find the next item in this line after or under the cursor and jump to its match. inclusive motion.
Items can be:
([{}])          parenthesis or (curly/square) brackets
                (this can be changed with the
                'matchpairs' option)
/* */           start or end of C-style comment
#if, #ifdef, #else, #elif, #endif
                C preprocessor conditionals (when the
                cursor is on the # or no ([{
                following)
```

aioutecism commented:
The `%` behavior in this p-r is not very correct.
`Find the next item in this line after or under the cursor and jump to its match.`

DanEEStar commented:
Ah, ok. Now I see it. I missed the "after" part.
Or is there something else that I missed?

aioutecism commented:
Just the "after" part is missing.
Reuse existing text object classes seems difficult.

DanEEStar commented:
I added the support for the "after" part.
Added comment that C-style comments (`/* ... */`) and C-preprocessor conditionals are not supported for now.

DanEEStar commented:
Can you comment the adjustments I made to the pull request? Thanks in advance!

aioutecism commented on src/Motions/MatchPair.ts:
```diff
@@ -0,0 +1,71 @@
+import {MotionMatch} from '../Motions/Match';
+import {TextObjectBlock} from '../TextObjects/Block';
+
+enum MotionMatchDirection {NEXT, PREV};
```
This enum is not used.

aioutecism commented on src/Motions/MatchPair.ts:
```diff
@@ -0,0 +1,71 @@
+import {window, Position, Range} from 'vscode';
+import {Motion} from './Motion';
+import {MotionMatch} from '../Motions/Match';
```
Unused import.

aioutecism commented on src/Motions/MatchPair.ts:
```diff
@@ -0,0 +1,71 @@
+    
+    // TODO: C-style comments (/* */) and C preprocessor conditionals are not supported for now
+    
+    private static openingCharacterMap = {
```
I suggest we define an interface and use it as this property's type, like this:

```ts
interface MotionMatchPairMap {
    [key: string]: (args) => TextObject;
}

// ...
    private static openingCharacterMap: MotionMatchPairMap = {
// ...
```

aioutecism commented on src/Motions/MatchPair.ts:
```diff
@@ -0,0 +1,71 @@
+                }
+            } else {
+                const endRange = textObject.findEndRange(document, new Position(from.line, i));
+                if(endRange !== null) {
```
The coding conventions of this project are to add a space between `if/for/while...` and `(`, and keep `else {` on a new line.
It would be great if you cloud flow the format.
Thanks in advance.

aioutecism commented on src/Motions/MatchPair.ts:
```diff
@@ -0,0 +1,71 @@
+        const document = activeTextEditor.document;
+        const targetText = document.lineAt(from.line).text;
+        
+        for(let i = from.character; i < targetText.length; i += 1) {
```
Using `i += 1` is ok, but people always use `i++` in js/ts.

aioutecism commented on src/Motions/MatchPair.ts:
```diff
@@ -0,0 +1,71 @@
+            const textObject = MotionMatchPair.characterMap[currentCharacter]({isInclusive: true});
+            
+            if(MotionMatchPair.openingCharacters.indexOf(currentCharacter) < 0) {
+                const startRange = textObject.findStartRange(document, new Position(from.line, i));
```
It would be easier to understand if we rename `i` to `character`, so `new Position(from.line, i)` would be `new Position(form.line, character)`.

aioutecism commented on src/Motions/MatchPair.ts:
```diff
@@ -0,0 +1,71 @@
+                continue;
+            }
+            
+            const textObject = MotionMatchPair.characterMap[currentCharacter]({isInclusive: true});
```
`isInclusive` is not used because we are not calling `apply`.

aioutecism commented:
`%` works like a charm!
Thanks!

DanEEStar commented:
Thanks for your comments, it is helpful to have them! 
I tried to adapt the code according to them.
I will try catch the tslint errors before I commit in the future.

aioutecism commented:
New version published.

vavans commented:
Thank for the work !
